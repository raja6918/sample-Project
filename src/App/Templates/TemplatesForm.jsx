import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import InputMUI from '@material-ui/core/TextField';
import MUIFormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';

import Dialog from '../../components/Dialog/Form';
import AutoComplete from '../../components/Autocomplete/AutoComplete.jsx';
import ErrorMessage from '../../components/FormValidation/ErrorMessage';
import TextArea from '../../components/TextArea';
import { TEMPLATE_NAME_REGEX } from './constants';
import { evaluateRegex } from '../../utils/common';

const FormControl = styled(MUIFormControl)`
  width: 100%;
`;

const AddDialog = styled(Dialog)`
  & div:last-child form {
    width: 560px;
    max-width: 560px;
  }

  & div:last-child form > div:first-child > span:nth-child(1) {
    margin-bottom: 10px;
    display: block;
    color: rgba(0, 0, 0, 0.87);
  }

  & div:last-child form > div:first-child > div:nth-child(4) {
    margin-top: 16px;
    & .Select-menu {
      max-height: 140px;
    }
  }
  & div:last-child form > div:first-child > div:nth-child(5) {
    margin-bottom: 10px;
  }
  & .MuiInputAdornment-positionStart {
    margin-right: 0px;
  }
`;

const Input = styled(InputMUI)`
  & label {
    color: rgba(0, 0, 0, 0.87);
  }

  &.error > label {
    color: #d10000 !important;
  }
`;

class TemplatesForm extends Component {
  state = {
    templateName: '',
    sourceTemplate: '-1',
    categoryName: '',
    errors: {
      templateName: false,
    },
  };

  validateTempNameAndSetErrors = () => {
    const { templateName, errors } = this.state;
    if (templateName !== '') {
      const isValidRegex = evaluateRegex(TEMPLATE_NAME_REGEX, templateName);
      this.setState({ errors: { ...errors, templateName: !isValidRegex } });
    } else {
      this.setState({ errors: { ...errors, templateName: false } });
    }
  };

  getTemplateId = templateName => {
    let templateToFind = '-1';
    this.props.templateSuggestions.forEach(template => {
      if (template.label === templateName) templateToFind = template.value;
    });

    return templateToFind;
  };
  componentDidMount() {
    this.setState({ sourceTemplate: this.getTemplateId('Blank') });
  }

  handleChange = e => {
    const name = e.target.name;
    this.setState({ [name]: e.target.value }, () => {
      switch (name) {
        case 'templateName':
          this.validateTempNameAndSetErrors();
          break;
        default:
          return 0;
      }
    });
  };

  isValidForm = () => {
    return (
      this.state.templateName.trim() === '' ||
      this.state.sourceTemplate === null ||
      this.state.sourceTemplate === '-1'
    );
  };

  componentWillReceiveProps(nextProps) {
    if (!nextProps.open) {
      this.setState({
        templateName: '',
        sourceTemplate: this.getTemplateId('Blank'),
        categoryName: '',
      });
    }
  }

  onAutocompleteChange(name, value) {
    this.setState({ [name]: value });
  }

  render() {
    const { formId, t, templateSuggestions, categories, ...rest } = this.props;
    const { templateName, sourceTemplate, categoryName, errors } = this.state;

    return (
      <AddDialog
        title={t('TEMPLATES.formTitle')}
        formId={formId}
        disableSave={this.isValidForm() || errors.templateName}
        {...rest}
      >
        <span>{t('TEMPLATES.sourceTemplateTitle')}</span>
        <AutoComplete
          id={'sourceTemplate'}
          name={'sourceTemplate'}
          placeholder={t('TEMPLATES.sourceTemplate')}
          suggestions={templateSuggestions}
          required
          value={sourceTemplate}
          defaultValue={this.getTemplateId('Blank')}
          t={t}
          onChange={value => this.onAutocompleteChange('sourceTemplate', value)}
        />
        <Input
          inputProps={{
            name: 'templateName',
            maxLength: 50,
            required: true,
          }}
          id="templateName"
          label={t('TEMPLATES.templateName')}
          margin="normal"
          required
          fullWidth
          value={templateName}
          onChange={this.handleChange}
          color="primary"
          className={errors.templateName ? 'error' : ''}
        />
        <ErrorMessage
          error={errors.templateName}
          isVisible={errors.templateName}
          message={t(`ERRORS.TEMPLATES.name`)}
        />
        <FormControl>
          <InputLabel htmlFor="category">{t('CATEGORY.label')}</InputLabel>
          <AutoComplete
            id={'category'}
            name={'category'}
            suggestions={categories}
            create={true}
            createLabel={t('CATEGORY.create')}
            t={t}
            maxLength={50}
            onChange={value => this.onAutocompleteChange('categoryName', value)}
            value={categoryName}
          />
        </FormControl>
        <TextArea
          inputProps={{
            name: 'description',
            maxLength: 1000,
          }}
          id="description"
          label={t('TEMPLATES.description')}
          margin="normal"
          style={{ display: 'block' }}
          fullWidth
          multiline
        />
      </AddDialog>
    );
  }
}
TemplatesForm.propTypes = {
  open: PropTypes.bool.isRequired,
  handleCancel: PropTypes.func.isRequired,
  handleOk: PropTypes.func.isRequired,
  formId: PropTypes.string.isRequired,
  t: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  okButton: PropTypes.string,
  cancelButton: PropTypes.string,
  templateSuggestions: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
    })
  ).isRequired,
  categories: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    })
  ).isRequired,
};

TemplatesForm.defaultProps = {
  okButton: '',
  cancelButton: '',
};

export default TemplatesForm;
