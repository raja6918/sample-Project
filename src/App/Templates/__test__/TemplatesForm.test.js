import React from 'react';
import { shallow } from 'enzyme';

import TemplatesForm from '../TemplatesForm';

const TemplatesSuggestions = [
  { value: 1, label: 'Template 1' },
  { value: 2, label: 'Template 2' },
  { value: 3, label: 'Template 3' },
  { value: 4, label: 'Template 4' },
];

const categories = [
  { value: 'Category 1', label: 'Category 1' },
  { value: 'Category 2', label: 'Category 2' },
  { value: 'Category 3', label: 'Category 3' },
];
test.skip('Form component renders', () => {
  const wrapper = shallow(
    <TemplatesForm
      open={false}
      handleCancel={jest.fn()}
      handleOk={jest.fn()}
      formId={'templateForm'}
      t={jest.fn()}
      onClose={jest.fn()}
      okButton={'CREATE'}
      cancelButton={'CANCEL'}
      templateSuggestions={TemplatesSuggestions}
      categories={categories}
    />
  );

  expect(wrapper).toMatchSnapshot();
});
