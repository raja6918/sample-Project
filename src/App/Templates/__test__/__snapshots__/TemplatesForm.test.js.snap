// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders 1`] = `
<TemplatesForm__AddDialog
  cancelButton="CANCEL"
  disableSave={true}
  formId="templateForm"
  handleCancel={[MockFunction]}
  handleOk={[MockFunction]}
  okButton="CREATE"
  onClose={[MockFunction]}
  open={false}
>
  <span />
  <AutoComplete
    adornmentIcon=""
    create={false}
    createLabel=""
    defaultValue="-1"
    disabled={false}
    id="sourceTemplate"
    maxLength={1000}
    name="sourceTemplate"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    placeholder=""
    required={true}
    setShrink={[Function]}
    style={Object {}}
    suggestions={
      Array [
        Object {
          "label": "Template 1",
          "value": 1,
        },
        Object {
          "label": "Template 2",
          "value": 2,
        },
        Object {
          "label": "Template 3",
          "value": 3,
        },
        Object {
          "label": "Template 4",
          "value": 4,
        },
      ]
    }
    t={
      [MockFunction] {
        "calls": Array [
          Array [
            "TEMPLATES.formTitle",
          ],
          Array [
            "TEMPLATES.sourceTemplateTitle",
          ],
          Array [
            "TEMPLATES.sourceTemplate",
          ],
          Array [
            "TEMPLATES.templateName",
          ],
          Array [
            "ERRORS.TEMPLATES.name",
          ],
          Array [
            "CATEGORY.label",
          ],
          Array [
            "CATEGORY.create",
          ],
          Array [
            "TEMPLATES.description",
          ],
          Array [
            "TEMPLATES.formTitle",
          ],
          Array [
            "TEMPLATES.sourceTemplateTitle",
          ],
          Array [
            "TEMPLATES.sourceTemplate",
          ],
          Array [
            "TEMPLATES.templateName",
          ],
          Array [
            "ERRORS.TEMPLATES.name",
          ],
          Array [
            "CATEGORY.label",
          ],
          Array [
            "CATEGORY.create",
          ],
          Array [
            "TEMPLATES.description",
          ],
        ],
      }
    }
    value="-1"
  />
  <TemplatesForm__Input
    className=""
    color="primary"
    fullWidth={true}
    id="templateName"
    inputProps={
      Object {
        "maxLength": 50,
        "name": "templateName",
        "required": true,
      }
    }
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <ErrorMessage
    error={false}
    isVisible={false}
    message=""
  />
  <TemplatesForm__FormControl>
    <WithStyles(ForwardRef(InputLabel))
      htmlFor="category"
    />
    <AutoComplete
      adornmentIcon=""
      create={true}
      createLabel=""
      disabled={false}
      id="category"
      maxLength={50}
      name="category"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      placeholder=""
      required={false}
      setShrink={[Function]}
      style={Object {}}
      suggestions={
        Array [
          Object {
            "label": "Category 1",
            "value": "Category 1",
          },
          Object {
            "label": "Category 2",
            "value": "Category 2",
          },
          Object {
            "label": "Category 3",
            "value": "Category 3",
          },
        ]
      }
      t={
        [MockFunction] {
          "calls": Array [
            Array [
              "TEMPLATES.formTitle",
            ],
            Array [
              "TEMPLATES.sourceTemplateTitle",
            ],
            Array [
              "TEMPLATES.sourceTemplate",
            ],
            Array [
              "TEMPLATES.templateName",
            ],
            Array [
              "ERRORS.TEMPLATES.name",
            ],
            Array [
              "CATEGORY.label",
            ],
            Array [
              "CATEGORY.create",
            ],
            Array [
              "TEMPLATES.description",
            ],
            Array [
              "TEMPLATES.formTitle",
            ],
            Array [
              "TEMPLATES.sourceTemplateTitle",
            ],
            Array [
              "TEMPLATES.sourceTemplate",
            ],
            Array [
              "TEMPLATES.templateName",
            ],
            Array [
              "ERRORS.TEMPLATES.name",
            ],
            Array [
              "CATEGORY.label",
            ],
            Array [
              "CATEGORY.create",
            ],
            Array [
              "TEMPLATES.description",
            ],
          ],
        }
      }
      value=""
    />
  </TemplatesForm__FormControl>
  <TextArea
    fullWidth={true}
    id="description"
    inputProps={
      Object {
        "maxLength": 1000,
        "name": "description",
      }
    }
    margin="normal"
    multiline={true}
    style={
      Object {
        "display": "block",
      }
    }
  />
</TemplatesForm__AddDialog>
`;
